spanner/client.go

```diff
@@ -669,6 +672,193 @@ func (c *Client) Apply(ctx context.Context, ms []*Mutation, opts ...ApplyOption)
+	for {
+		// Stream finished or in error state.
+		if r.err != nil {
+			return nil, r.err
+		}
+
+		// RPC not made yet.
+		if r.stream == nil {
+			r.stream, r.err = r.rpc(r.ctx)
+			continue
+		}
+
+		// Read from the stream.
+		var response *sppb.BatchWriteResponse
+		response, r.err = r.stream.Recv()
+
+		// Return an item.
+		if r.err == nil {
+			r.dataReceived = true
+			return response, nil
+		}
+
+		// Stream finished.
+		if r.err == io.EOF {
+			r.err = iterator.Done
+			return nil, r.err
+		}
+
+		// Retry request on session not found error only if no data has been received before.
+		if !r.dataReceived && r.replaceSession != nil && isSessionNotFoundError(r.err) {
+			r.err = r.replaceSession(r.ctx)
+			r.stream = nil
+		}
+	}
+}
+
+// Stop terminates the iteration. It should be called after you finish using the
+// iterator.
+func (r *BatchWriteResponseIterator) Stop() {
+	if r.stream != nil {
+		err := r.err
+		if err == iterator.Done {
+			err = nil
+		}
+		defer trace.EndSpan(r.ctx, err)
+	}
+	if r.cancel != nil {
+		r.cancel()
+		r.cancel = nil
+	}
+	if r.release != nil {
+		r.release(r.err)
+		r.release = nil
+	}
+	if r.err == nil {
+		r.err = spannerErrorf(codes.FailedPrecondition, "Next called after Stop")
+	}
+}
+
+// Do calls the provided function once in sequence for each item in the
+// iteration. If the function returns a non-nil error, Do immediately returns
+// that error.
+//
+// If there are no items in the iterator, Do will return nil without calling the
+// provided function.
+//
+// Do always calls Stop on the iterator.
+func (r *BatchWriteResponseIterator) Do(f func(r *sppb.BatchWriteResponse) error) error {
+	defer r.Stop()
+	for {
+		row, err := r.Next()
+		switch err {
+		case iterator.Done:
+			return nil
+		case nil:
+			if err = f(row); err != nil {
+				return err
+			}
+		default:
+			return err
+		}
+	}
+}
+
+// BatchWrite applies a list of mutation groups in a collection of efficient
+// transactions. The mutation groups are applied non-atomically in an
+// unspecified order and thus, they must be independent of each other. Partial
+// failure is possible, i.e., some mutation groups may have been applied
+// successfully, while some may have failed. The results of individual batches
+// are streamed into the response as the batches are applied.
+//
+// BatchWrite requests are not replay protected, meaning that each mutation
+// group may be applied more than once. Replays of non-idempotent mutations
+// may have undesirable effects. For example, replays of an insert mutation
+// may produce an already exists error or if you use generated or commit
+// timestamp-based keys, it may result in additional rows being added to the
+// mutation's table. We recommend structuring your mutation groups to be
+// idempotent to avoid this issue.
+func (c *Client) BatchWrite(ctx context.Context, mgs []*MutationGroup, opts ...ApplyOption) *BatchWriteResponseIterator {
```

Comment:
`ApplyOption` would not be suitable here since it is an argument tailored to `Apply` method. Additionally, when new options need to be introduced for BatchWrite in future it will add the change to `ApplyOptions` which might not be valid for `Apply` method and might confuse users.
IMO, introducing new `BatchWriteOption` would be more appropriate here.

I am trying to explore [Go generics](https://go.dev/doc/tutorial/generics) to avoid duplicate code for `ApplyOption` and `BatchWriteOption`, but the solution seems to be tricky. 
Please let me know if you have some better way to do this.

Reply:
@harshachinta - Thanks for the feedback. This makes sense.

How about we expose a new struct `BatchWriteOptions`, similar to `QueryOptions` -
```go
struct BatchWriteOptions {
  Priority sppb.RequestOptions_Priority,
  TransactionTag String,
  // future options go here
}
```

The interface would then become - 
```go
func (c *Client) BatchWrite(ctx context.Context, mgs []*MutationGroup, opts BatchWriteOptions) *BatchWriteResponseIterator {}
```

The `Client` struct would contain `bwo` as a field and there'd be a new `func (bwo BatchWriteOptions) merge(opts BatchWriteOptions) BatchWriteOptions {}` function.

If the above looks good to you, I'll make this change.

**Aside (not related to this PR) -**
All of the different `options` constructs seem to contain fields from the more generic `sppb.RequestOptions` struct along with other request specific options. I suppose, `sppb.RequestOptions` could be modelled directly and the specific request options structs could contain this generic options struct.

Reply:
That looks like one possible option. To collate, there are 2 ways to go about this,
1. As mentioned above, introduce
```
struct BatchWriteOptions {
  Priority sppb.RequestOptions_Priority
  TransactionTag String
  // future options go here
}
```
which will have the following API definitions,
```
1. func (c *Client) BatchWrite(ctx context.Context, mgs []*MutationGroup) *BatchWriteResponseIterator {}
2. func (c *Client) BatchWriteWithOptions(ctx context.Context, mgs []*MutationGroup, opts BatchWriteOptions) *BatchWriteResponseIterator {}
```

2. Second way is to do the following,
Add a new type `APIOption` that can be used across all new APIs
```
type APIOption func(T interface{})

// Example of one option that customer can use
func TransactionTagOption(tag string) APIOption {
	return func(o interface{}) {
		switch o.(type) {
		case *batchWriteOption:
			o.(*batchWriteOption).transactionTag = tag
                // can be extended to new API in future
		}
	}
}
```
and create a `batchWriteOption` type and use it inside the `BatchWrite` method.
```
type batchWriteOption struct {
	transactionTag string
	priority sppb.RequestOptions_Priority
}

func (c *Client) BatchWrite(ctx context.Context, mgs []*MutationGroup, opts ...APIOption) *BatchWriteResponseIterator {
	bo := &batchWriteOption{}
	for _, opt := range opts {
		opt(bo)
	}
}
```
This way any new APIs/methods in future can have its options under APIOption.

@olavloite @rahul2393 Can you please suggest if any of these two ways would be good or suggest any other possible way?

Reply:
@harshachinta - Went ahead and implemented option 1 in 5b9ce8a11ca9efc3976bb5ac579c132daf749c7d. PTAL if/when there's agreement on the approach.

Reply:
Option 1 seems to be a good choice here. Lets go ahead with that approach.